plugins/usbpro/OpenDeckDevice.cpp

```diff
@@ -0,0 +1,239 @@
+/*
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Library General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * OpenDeckDevice.cpp
+ * An OpenDeck Device
+ * Copyright (C) 2022 Igor Petrovic
+ *
+ * This device creates a single OpenDeck output port
+ */
+
+#include <iomanip>
+#include <ios>
+#include <iostream>
+#include <string>
+
+#include "common/rpc/RpcController.h"
+#include "ola/Callback.h"
+#include "ola/Constants.h"
+#include "ola/Logging.h"
+#include "ola/strings/Format.h"
+#include "plugins/usbpro/OpenDeckDevice.h"
+#include "plugins/usbpro/UsbProWidgetDetector.h"
+
+namespace ola {
+namespace plugin {
+namespace usbpro {
+
+using ola::plugin::usbpro::Request;
+using ola::plugin::usbpro::Reply;
+using ola::rpc::RpcController;
+using ola::utils::SplitUInt16;
+using ola::strings::ToHex;
+using std::string;
+
+/*
+ * Create a new device
+ * @param owner  the plugin that owns this device
+ * @param name  the device name
+ * @param dev_path  path to the OpenDeck widget
+ */
+OpenDeckDevice::OpenDeckDevice(ola::PluginAdaptor *plugin_adaptor,
+                               ola::AbstractPlugin *owner,
+                               const string &name,
+                               OpenDeckWidget *widget,
+                               OLA_UNUSED uint16_t esta_id,
+                               OLA_UNUSED uint16_t device_id,
+                               uint32_t serial,
+                               uint16_t firmware_version,
+                               unsigned int fps_limit):
+    UsbSerialDevice(owner, name, widget),
+    m_opendeck_widget(widget),
+    m_serial(),
+    m_got_parameters(false),
+    m_break_time(0),
+    m_mab_time(0),
+    m_rate(0) {
+  // Serial number is stored as series of 4 bytes.
+  // Display the serial in hex format so that it
+  // matches (partially) with USB serial number.
```

Comment:
Even better, I'd not even checked how they behave! Does this also match representations you use within your system; that's probably more important knowing you're talking about the same device both ends!

Reply:
Yeah it does now, although I had the wrong byte order on the device which is fixed now.

OLA:
`olad/plugin_api/DeviceManager.cpp:105: Installed device: Shantea Controls - discovery:5-0x002e001d`

dmesg:
```
[363290.808545] usb 1-2.1: Product: OpenDeck | discovery
[363290.808546] usb 1-2.1: Manufacturer: Shantea Controls
[363290.808547] usb 1-2.1: SerialNumber: 002E001D574850012032364B
[363290.822563] cdc_acm 1-2.1:1.0: ttyACM1: USB ACM device
```

Reply:
Additional change to truly match (no 0x prefix and uppercase now)

`olad/plugin_api/DeviceManager.cpp:105: Installed device: Shantea Controls - discovery:5-002E001D`

Reply:
I assume there's no benefit to taking one end of the serial or the other from your device? e.g. is the lower end more likely to be different between two devices produced around the same time?

Reply:
They're both equally random.